### PR TITLE
Add table for grad modes in documentation

### LINKED ISSUE #104461: Add a table to the grad modes documentation to help readability
Add a table to the grad modes documentation to help readability

cc @svekars @carljparker @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by davidradl**
Fixes #104461

### CODE REVIEW

File: docs/source/notes/autograd.rst [hunk 1]
```diff
@@ -170,6 +170,32 @@ processed by autograd internally: default mode (grad mode), no-grad mode,
 and inference mode, all of which can be togglable via context managers and
 decorators.
 
+
+.. list-table::
+   :widths: 50 50 50 50 50
+   :header-rows: 1
+
+   * - Mode
+     - required-grad takes effect
```
Comment (mikaylagawarecki) [inline docs/source/notes/autograd.rst:179]: nit: 
```suggestion
     - ``requires_grad`` takes effect
```

**Review by svekars — APPROVED** [review]

File: docs/source/notes/autograd.rst [hunk 2]
```diff
@@ -170,6 +170,32 @@ processed by autograd internally: default mode (grad mode), no-grad mode,
...
+     - Recorded in backwards graph
```
Comment (soulitzer) [inline docs/source/notes/autograd.rst:180]: Operations are recorded in the backwards graph

File: docs/source/notes/autograd.rst [hunk 3]
```diff
@@ -170,6 +170,32 @@ processed by autograd internally: default mode (grad mode), no-grad mode,
...
+     - Output tensor can be used in grad-mode later
```
Comment (soulitzer) [inline docs/source/notes/autograd.rst:181]: Tensors created while the mode is enabled can be used in grad mode later

File: docs/source/notes/autograd.rst (hunk 1 above)
Comment (soulitzer) [reply docs/source/notes/autograd.rst:179]: autograd respects the ``requires_grad`` of operator inputs

File: docs/source/notes/autograd.rst [hunk 4]
```diff
@@ -170,6 +170,32 @@ processed by autograd internally: default mode (grad mode), no-grad mode,
...
+   * - grad (default)
+     - ✓
+     - ✓
+     - ✓
+     - Forward pass
+   * - no-grad
+     -
+     -
+     - ✓
+     - Optimizer
```
Comment (soulitzer) [inline docs/source/notes/autograd.rst:192]: optimizer updates

**Review by soulitzer — CHANGES_REQUESTED** [review]
Thanks for the addition, but I'm not sure a table would make things much clearer here:

- Some of the items would be quite long if we were to make them more precise (I added some suggestions)
- The first and second columns are saying very similar things (why do we need to have both?)
- People are mainly confused about no-grad mode vs inference mode right? Do we need grad mode in the table as well?

File: docs/source/notes/autograd.rst (hunk 1 above)
Comment (davidradl) [reply docs/source/notes/autograd.rst]: changed

**Comment by davidradl** [comment]
@soulitzer thank for your feedback - your descriptions are clearer and more accurate.

**Review by soulitzer — COMMENTED** [review]
Thanks for the update.

I think the table would be even more helpful if we focus on the difference between no-grad and inference mode (and also skip the "autograd respects require_grad" since thats more of an implementation detail of how operations are excluded from the graph).

No grad mode:
- excludes operations from being recorded in backward graph (yes)
- skips additional autograd tracking overhead (no)
- tensors created in this mode can be used with autograd later (yes)
- example: optimizer updates

Inference mode:
- excludes operations from being included in backward graph (yes)
- skips additional autograd tracking overhead (yes)
- tensors created in this mode can be used with autograd later (no)
- example: data processing, model evaluation

**Comment by davidradl** [comment]
@soulitzer Thank you again for the feedback. I agree with removing` autograd respects require_grad`.

I think it would be helpful to include grad mode, as this can then be contrasted in the table with the other 2 modes. 

I am also thinking that the column heading should be positive, so it is simpler English to avoid the double negative. 

-  `includes operations in the backward graph (no)` instead of `excludes operations from being included in backward graph (yes)`
- `Additional autograd tracking overhead (yes)` instead of `skips additional autograd tracking overhead (no)`

If you have preferred positive workings I cam happy to amend .

I wanted to run this by you before submitting any new changes.

**Comment by soulitzer** [comment]
Keeping grad mode row sounds okay, but let's rename to 'default' instead so we don't mislead users into thinking that grad mode is something they would need to enable.

I don't think `excludes operations from being included in backward graph (yes)` and `skips additional autograd tracking overhead (no)` are truly negative wording in the sense that this table is supposed to be about the benefits you would get from enabling these alternative modes, so users would be looking for the options to be worded in this way.

`Additional autograd tracking overhead (yes)` doesn't quite work because it begs the question of additional relative to what. And just `autograd tracking` doesn't quite work because then I'd be wondering what the difference is that with just recording operations in the graph.

**Comment by davidradl** [comment]
> Keeping grad mode row sounds okay, but let's rename to 'default' instead so we don't mislead users into thinking that grad mode is something they would need to enable.
> 
> I don't think `excludes operations from being included in backward graph (yes)` and `skips additional autograd tracking overhead (no)` are truly negative wording in the sense that this table is supposed to be about the benefits you would get from enabling these alternative modes, so users would be looking for the options to be worded in this way.
> 
> I was also struggling to find a clear "positive" wording here without having to go into more implementation details which I'd like to avoid. `Additional autograd tracking overhead (yes)` doesn't quite work because it begs the question of additional relative to what. And just `autograd tracking` doesn't quite work because then I'd be wondering what the difference is that with just recording operations in the graph.
> 
> `skips additional autograd tracking overhead (no)` -> `skips more autograd tracking overhead (no)` might be slightly clearer

@soulitzer makes sense, I will put up a changed table tomorrow - thank you .

**Comment by soulitzer** [comment]
Looks good otherwise, thanks for the updates and the discussion

File: third_party/cutlass
[the remark this answers is not in the document]
Comment (davidradl) [reply third_party/cutlass:1]: Hi @soulitzer, this is a merge commit. I goto [https://github.com/pytorch/pytorch/pull/104462/files#diff-22f3d369e722a09f57144cd597deed394606b8bc4429247d44d0ecaa001aea5e](https://github.com/pytorch/pytorch/pull/104462/files#diff-22f3d369e722a09f57144cd597deed394606b8bc4429247d44d0ecaa001aea5e) and check the diffs the files are the same as base. I think the +es -es refer to the original commit. I don't think I am regressing/ introducing anything, as there is no diff for any of the files. 

I am not sure how you would like me to handle this. We can:
1) Leave as is if you agree with me that this is benign
2) I could force my local branch back to current add my change then force push that change up. 
3) Submit another pr with only my change without any merge commits.

If you have any doubts on 1) then 3) would be simplest.

[the remark this answers is not in the document]
Comment (soulitzer) [reply third_party/cutlass:1]: new PR sounds good!

**Comment by davidradl** [comment]
closing and pursuing fix in pr https://github.com/pytorch/pytorch/pull/104774 as agreed

**Pull request closed without merging**